_episodes/03-Unit_Testing.md

```diff
@@ -0,0 +1,147 @@
+---
+
+Unit tests are so called because they exercise the functionality of the code by
+interrogating individual functions and methods. Functions and methods can often
+be considered the atomic units of software because they are indivisible.
+However, what is considered to be the smallest code _unit_ is subjective. The
+body of a function can be long or short, and shorter functions are arguably
+more unit-like than long ones.
+
+Thus what reasonably constitutes a code unit typically varies from project to
+project and language to language.  A good guideline is that if the code cannot
+be made any simpler logically (you cannot split apart the addition operator) or
+practically (a function is self-contained and well defined), then it is a unit.
+
+> ## Functions are Like Paragraphs
+>
+> Recall that humans can only hold a few ideas in our heads at once. Paragraphs
+> in books, for example, become unwieldy after a few lines. Functions, generally,
+> shouldn't be longer than paragraphs.
+> Robert C. Martin, the author of "Clean Code" said : "The first rule of
+> functions is that _they should be small_. The second rule of functions is that
+> _they should be smaller than that_."
+{: .callout}
+
+The desire to unit test code often has the effect of encouraging both the
+code and the tests to be as small, well-defined, and modular as possible.
+In Python, unit tests typically take the form of test functions that call and make
+assertions about methods and functions in the code base.  To run these test
+functions, a test framework is often required to collect them together. For
+now, we'll write some tests for the mean function and simply run them
+individually to see whether they fail. In the next session, we'll use a test
+framework to collect and run them.
+
+## Unit Tests Are Just Functions
+
+Unit tests are typically made of three pieces, some set-up, a number of
+assertions, and some tear-down. Set-up can be as simple as initializing the
+input values or as complex as creating and initializing concrete instances of a
+class. Ultimately, the test occurs when an assertion is made, comparing the
+observed and expected values. For example, let us test that our mean function
+successfully calculates the known value for a simple list.
+
+Before running the next code, save your `mean` function to a file called mean.py in the working directory.
+
+You can use this code to save to file:
+
+~~~
+def mean(num_list):
+    try:
+        return sum(num_list)/len(num_list)
+    except ZeroDivisionError :
+        return 0
+    except TypeError as detail :
+        msg = "The algebraic mean of an non-numerical list is undefined.\
+               Please provide a list of numbers."
+        raise TypeError(detail.__str__() + "\n" +  msg)
+~~~
+{: .python}
+
+Now, run the following code:
+
+~~~
+from mean import mean
+def test_ints():
+    num_list = [1, 2, 3, 4, 5]
+    obs = mean(num_list)
+    exp = 3
+    assert obs == exp
```

Comment:
since we don't introduce setup/teardown, this seems needlessly wordy just do `approx(3) == mean(range(1, 6))`